File: internal/models/user_state.go
package models

import (
	"github.com/google/uuid"
	"github.com/juevigrace/diva-server/internal/models/responses"
	"github.com/juevigrace/diva-server/storage"
)

type UserState struct {
	Verified     bool
	Status       UserStatus
	LastActiveAt int64
	UpdatedAt    int64
}

func (us *UserState) Response() *responses.UserStateResponse {
	return &responses.UserStateResponse{
		Verified:     us.Verified,
		Status:       us.Status.String(),
		LastActiveAt: us.LastActiveAt,
		UpdatedAt:    us.UpdatedAt,
	}
}

func (us *UserState) DBCreate(userID uuid.UUID) *storage.CreateUserStateParams {
	return &storage.CreateUserStateParams{
		UserID:   userID,
		Verified: us.Verified,
		Status:   us.Status.ToDB(),
	}
}

func UserStateFromDB(row *storage.DivaUserState) *UserState {
	return &UserState{
		Verified:     row.Verified,
		Status:       UserStatusFromDB(row.Status),
		LastActiveAt: row.LastActiveAt,
		UpdatedAt:    row.UpdatedAt,
	}
}

File: client/packages/shared/app-lib/diva-lib-models-core/src/commonMain/kotlin/io/github/juevigrace/diva/lib/models/user/state/UserState.kt
package io.github.juevigrace.diva.lib.models.user.state

import io.github.juevigrace.diva.core.Option
import io.github.juevigrace.diva.lib.models.api.user.state.UserStateResponse
import io.github.juevigrace.diva.lib.models.user.UserStatus
import io.github.juevigrace.diva.lib.models.user.safeUserStatus
import kotlin.time.Instant
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

@OptIn(ExperimentalUuidApi::class)
data class UserState(
    val userId: Uuid = Uuid.NIL,
    val verified: Boolean = false,
    val status: UserStatus = UserStatus.ACTIVE,
    val lastActiveAt: Option<Instant> = Option.None,
    val updatedAt: Option<Instant> = Option.None,
) {
    companion object {
        fun fromResponse(response: UserStateResponse): UserState {
            return UserState(
                userId = Uuid.NIL,
                verified = response.verified,
                status = safeUserStatus(response.status),
                lastActiveAt = Option.of(response.lastActiveAt.let { Instant.fromEpochMilliseconds(it) }),
                updatedAt = Option.of(response.updatedAt.let { Instant.fromEpochMilliseconds(it) }),
            )
        }
    }
}
